flake.nix

```diff
@@ -0,0 +1,184 @@
+          tag = "example";
+          # mount and umount are required for csi-driver functions.
+          copyToRoot = pkgs.buildEnv {
+            name = "mount-bin";
+            pathsToLink = [ "/bin" ];
+            paths = [ pkgs.mount pkgs.umount ];
+          };
+          config = {
+            Description = "cert-manager CSI Driver";
+            Entrypoint = [ "${cert-manager-csi}/bin/cert-manager-csi" ];
+          };
+        };
+
+        cert-manager-helm-chart = pkgs.fetchurl {
+          url = "https://charts.jetstack.io/charts/cert-manager-v${e2e-cert-manager-version}.tgz";
+          sha256 = "sha256-Ricxd7XKaTCvgEfNUr8LeEyEgqwELngS0IZDwomGktU=";
+        };
+
+        kind-node-image = pkgs.dockerTools.pullImage{
+          imageName = "kindest/node";
+          imageDigest = e2e-kind-node-digest;
+          sha256 = e2e-kind-node-sha256;
+          finalImageTag = "v${e2e-kind-node-version}";
+          finalImageName = "kindest/node";
+        };
+
+        cert-manager-controller-image = pkgs.dockerTools.pullImage{
+          imageName = "quay.io/jetstack/cert-manager-controller";
+          imageDigest = e2e-cert-manager-controller-digest;
+          sha256 = e2e-cert-manager-controller-sha256;
+          finalImageTag = e2e-cert-manager-version;
+          finalImageName = "quay.io/jetstack/cert-manager-controller";
+        };
+
+        cert-manager-webhook-image = pkgs.dockerTools.pullImage{
+          imageName = "quay.io/jetstack/cert-manager-webhook";
+          imageDigest = e2e-cert-manager-webhook-digest;
+          sha256 = e2e-cert-manager-webhook-sha256;
+          finalImageTag = e2e-cert-manager-version;
+          finalImageName = "quay.io/jetstack/cert-manager-webhook";
+        };
+
+        cert-manager-cainjector-image = pkgs.dockerTools.pullImage{
+          imageName = "quay.io/jetstack/cert-manager-cainjector";
+          imageDigest = e2e-cert-manager-cainjector-digest;
+          sha256 = e2e-cert-manager-cainjector-sha256;
+          finalImageTag = e2e-cert-manager-version;
+          finalImageName = "quay.io/jetstack/cert-manager-cainjector";
+        };
+
+        cert-manager-ctl-image = pkgs.dockerTools.pullImage{
+          imageName = "quay.io/jetstack/cert-manager-ctl";
+          imageDigest = e2e-cert-manager-ctl-digest;
+          sha256 = e2e-cert-manager-ctl-sha256;
+          finalImageTag = e2e-cert-manager-version;
+          finalImageName = "quay.io/jetstack/cert-manager-ctl";
+        };
+
+        csi-node-driver-registrar-image = pkgs.dockerTools.pullImage{
+          imageName = "quay.io/k8scsi/csi-node-driver-registrar";
+          imageDigest = csi-node-driver-registrar-digest;
+          sha256 = csi-node-driver-registrar-sha256;
+          finalImageTag = "v${csi-node-driver-registrar-version}";
+          finalImageName = "quay.io/k8scsi/csi-node-driver-registrar";
+        };
+
+        busybox-image = pkgs.dockerTools.pullImage{
+          imageName = "busybox";
```

Comment:
What's the reasoning to switching to busybox and manually copying in binaries from the host/build environment rather than using alpine and `apk` to install the required tools? Does nix allow running arbitrary commands whilst building docker images or does it prevent it to ensure a hermetic build?

Reply:
We are actually pulling busybox here as that is the example application we use in `/deploy/example/example-app.yaml`. This basically means we can load the image in, rather than it being downloaded every time inside kind. This speeds up tests with a warm cache.

AFAIK nix also does not allow you to run arbitrary commands when building images.